cloudstack_plugin/virtual_machine.py

```diff
@@ -602,6 +608,63 @@ def disconnect_floating_ip(ctx, **kwargs):
     return True
 
 
+@operation
+def attach_volume(ctx, **kwargs):
+    """ Attach a volume to virtual machine.
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    server_id = ctx.source.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    volume_name = ctx.source.instance.runtime_properties.get(
+        CLOUDSTACK_ID_PROPERTY, None)
+
+    external_id = ctx.target.instance.runtime_properties.get(
+        CLOUDSTACK_ID_PROPERTY, None)
+
+    if external_id is not None:
+        volume_id = ctx.target.instance.runtime_properties[
+            CLOUDSTACK_ID_PROPERTY]
+
+        server = get_vm_by_id(ctx, cloud_driver, server_id)
+        volume = get_volume_by_id(cloud_driver, volume_id)
+
+        if server is None:
+            raise NonRecoverableError('Server with id {0} not found'
+                                      .format(server_id))
+
+        if volume is None:
+            raise NonRecoverableError('Volume with id {0} not found'
+                                      .format(volume_id))
+
+        cloud_driver.attach_volume(node=server,
+                                   volume=volume)
+
+    else:
+        ctx.logger.info('Volume {0} already exists and use of external '
```

Comment:
i dont understand this. why if external_id is None this means that the volume already exists? isnt it the other way around? and what does it have to do with attachement?